lib/assert.js

```diff
@@ -931,6 +933,83 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class CallTracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_CALLS_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      if (context.actual > context.exact) {
+        // If call count exceeds exact throw an error
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+
+        throw new AssertionError({
+          actual: context.actual,
+          expected: context.exact,
+          message,
+          operator: context.name,
+        });
+      }
```

Comment:
What is the purpose of this block of code?

Reply:
The literal purpose? When the function has been called exact times it's removed from the callChecks array. If it's called more than exact times an error is thrown. I could have it be added back into the callChecks array, if that would be preferable

Reply:
Yea, I think we should only enforce the checks from `verify()`.